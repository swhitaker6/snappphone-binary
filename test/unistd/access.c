/*
 * Copyright 2011 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <emscripten.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <sys/stat.h>
#include <unistd.h>

int main() {
  EM_ASM(
    FS.mkdir('working');
#if NODEFS
    FS.mount(NODEFS, { root: '.' }, 'working');
#endif
    FS.chdir('working');
    FS.writeFile('forbidden', ""); FS.chmod('forbidden', 0o000);
    FS.writeFile('readable',  ""); FS.chmod('readable',  0o444);
    FS.writeFile('writeable', ""); FS.chmod('writeable', 0o222);
    FS.writeFile('allaccess', ""); FS.chmod('allaccess', 0o777);
  );

  // Empty path checks #9136 fix
  char* files[] = {"readable", "writeable",
                   "allaccess", "forbidden", "nonexistent", ""};
  for (int i = 0; i < sizeof files / sizeof files[0]; i++) {
    printf("F_OK(%s): %d\n", files[i], access(files[i], F_OK));
    printf("errno: %d\n", errno);
    errno = 0;
    printf("R_OK(%s): %d\n", files[i], access(files[i], R_OK));
    printf("errno: %d\n", errno);
    errno = 0;
    printf("X_OK(%s): %d\n", files[i], access(files[i], X_OK));
    printf("errno: %d\n", errno);
    errno = 0;
    printf("W_OK(%s): %d\n", files[i], access(files[i], W_OK));
    printf("errno: %d\n", errno);
    errno = 0;
    printf("\n");
  }

  EM_ASM({FS.writeFile('filetorename',  'renametest');});

  int rename_ret = rename("filetorename", "renamedfile");
  assert(rename_ret == 0);

  errno = 0;
  printf("F_OK(%s): %d\n", "filetorename", access("filetorename", F_OK));
  printf("errno: %d\n", errno);
  errno = 0;
  printf("F_OK(%s): %d\n", "renamedfile", access("renamedfile", F_OK));
  printf("errno: %d\n", errno);

  // Same againt with faccessat
  errno = 0;
  printf("F_OK(%s): %d\n", "filetorename", faccessat(AT_FDCWD, "filetorename", F_OK, 0));
  printf("errno: %d\n", errno);
  errno = 0;
  printf("F_OK(%s): %d\n", "renamedfile", faccessat(AT_FDCWD, "renamedfile", F_OK, 0));
  printf("errno: %d\n", errno);

  // Restore full permissions on all created files so that python test runner rmtree
  // won't have problems on deleting the files. On Windows, calling shutil.rmtree()
  // will fail if any of the files are read-only.
  EM_ASM(
    FS.chmod('forbidden', 0o777);
    FS.chmod('readable',  0o777);
    FS.chmod('writeable', 0o777);
    FS.chmod('allaccess', 0o777);
  );

  return 0;
}
