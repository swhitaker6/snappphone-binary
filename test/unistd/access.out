F_OK(readable): 0
errno: 0
R_OK(readable): 0
errno: 0
X_OK(readable): -1
errno: 2
W_OK(readable): -1
errno: 2

F_OK(writeable): 0
errno: 0
R_OK(writeable): -1
errno: 2
X_OK(writeable): -1
errno: 2
W_OK(writeable): 0
errno: 0

F_OK(allaccess): 0
errno: 0
R_OK(allaccess): 0
errno: 0
X_OK(allaccess): 0
errno: 0
W_OK(allaccess): 0
errno: 0

F_OK(forbidden): 0
errno: 0
R_OK(forbidden): -1
errno: 2
X_OK(forbidden): -1
errno: 2
W_OK(forbidden): -1
errno: 2

F_OK(nonexistent): -1
errno: 44
R_OK(nonexistent): -1
errno: 44
X_OK(nonexistent): -1
errno: 44
W_OK(nonexistent): -1
errno: 44

F_OK(): -1
errno: 44
R_OK(): -1
errno: 44
X_OK(): -1
errno: 44
W_OK(): -1
errno: 44

F_OK(filetorename): -1
errno: 44
F_OK(renamedfile): 0
errno: 0
F_OK(filetorename): -1
errno: 44
F_OK(renamedfile): 0
errno: 0
